docs/guides/stream-intercept.mdx

```diff
@@ -0,0 +1,95 @@
+---
+slug: guides/stream-intercept
+title: SSR Stream Interception
+description: How to intercept SSR stream to inject content before it is sent to the client.
+---
+
+## Additional Content Injection in SSR
+
+In some cases, you might want to inject additional content into the SSR stream before it is sent
+to the client. This can be useful for injecting propperties onto the document root element for instanceof,
+like the `lang` attribute or a class name based on request details.
+
+### Step 1: Create a custom middleware for Waku
+
+Create a new file in your project, for example `src/middleware/streamInterceptor.ts`:
+
+```ts
+import type { Middleware } from "waku/config";
+
+// key below is probably best to have at a central config location to allow for it to be changed easily
+export const themeCookieKey = "waku-color-theme";
+
+const streamInterceptor: Middleware = () => {
+  return async (ctx, next) => {
+    const theme = ctx.context.theme;
+
+    await next();
+
+    ctx.res.headers ||= {};
+
+    const isDocument = ctx.res.headers["content-type"]?.includes("text/html");
+    const lang = "en-US";
+
+    if (isDocument) {
+      const documentContent = ctx.res.body;
+
+      const newDocumentContent = documentContent?.pipeThrough(
+        new TransformStream({
+          async transform(chunk, controller) {
+            const text = new TextDecoder().decode(chunk);
+            const newText = text.replace(
+              "<html>",
+              `<html lang="${lang}" ${theme == "dark" ? "class='dark'" : ""}>`,
+            );
+
+            controller.enqueue(new TextEncoder().encode(newText));
+          },
+        }),
+      );
+
+      ctx.res.body = newDocumentContent!;
+    }
+  };
+};
+
+export default streamInterceptor;
+```
+
+### Step 2: Register the middleware
+
+In your `waku.config.ts` file, import the middleware and add it to the `middlewares` array:
+
+```ts
+import streamInterceptor from "./src/middleware/streamInterceptor";
+
+const DO_NOT_BUNDLE = "";
+
+/** @type {import('waku/config').Config} */
+export default {
+  // ...
+  middlewares: (cmd: "dev" | "start") => [
+    import("./src/middleware/themePreload.js"), // + this is the one we just added
+    ...(cmd === "dev"
+      ? [
+          import(
+            /* @vite-ignore */ DO_NOT_BUNDLE + "waku/middleware/dev-server"
+          ),
+        ]
+      : []),
```

Comment:
Note: This is unavoidable for now. I wish we find a better way to simply the user config.